scenario/state.py

```diff
@@ -660,6 +667,17 @@ def get_filesystem(self, ctx: "Context") -> Path:
         """Simulated pebble filesystem in this context."""
         return ctx._get_container_root(self.name)
 
+    def get_exec(self, command_prefix: Sequence[str]):
+        """Get the Exec object from the container with the given command prefix."""
+        for exec in self.execs:
+            if exec.command_prefix == command_prefix:
+                return exec
+        raise KeyError(f"no exec found with command prefix {command_prefix}")
+
+    def _update_execs(self, execs: Iterable[Exec]):
+        # bypass frozen dataclass
+        object.__setattr__(self, "execs", frozenset(execs))
```

Comment:
A bit unfortunate, but maybe it's worth it, as opposed to changing a bunch of methods to return (modified) Self and forcing our users to adapt to the immutable API?

It would be more explicit, but maybe a bit confusing if charm writers are not Python experts.

Example how proper immutability would look to the user:

```py
proc = ...
proc, stdout, _ = proc.wait_output(...)
```

Reply:
The charmers do have to work with the (mostly) immutable objects, but they generally work at a higher level. If you wanted to make the container fail `can_connect`, for example:

```python
# test one
container = Container("foo", can_connect=True)
state_in = State(containers=[container])
state_out = ctx.run(container.pebble_ready_event, state_in)

# new test
container2 = dataclasses.replace(container, can_connect=False)
state_in2 = dataclasses.replace(state_in, containers=[container2])
state_out2 = ctx.run(container.pebble_ready_event, state_in2)
```

However, with the `proc.wait_output` code, that's not in the tests, that will be in the charm code itself, which knows nothing about Scenario. It thinks that it's working with a `ops.pebble.Client` (via the `ops.Model` wrapper).

The `object.__setattr__` is because we need to build up an output (mostly) immutable state that's a copy of the input state with whatever changes come from running the event.

It's definitely not ideal. The changes that Pietro is considering in #129 might address it somewhat. We might want to refactor the implementation at some point to avoid it (e.g. to have a private mutable state that's used during the event run and then 'frozen' when returned to the caller) - but that's definitely something for another day (and since it's unlikely to change the API, can wait until later, since we can do it in a minor release).